lib/matplotlib/category.py

```diff
@@ -1,123 +1,163 @@
+    if LooseVersion(np.__version__) < LooseVersion('1.7.0'):
+        if (isinstance(value, (six.text_type, np.unicode))):
+            value = value.encode('utf-8', 'ignore').decode('utf-8')
+    if isinstance(value, (np.bytes_, six.binary_type)):
+        value = value.decode(encoding='utf-8')
+    elif not isinstance(value, (np.str_, six.string_types)):
+        value = str(value)
+    return value
 
 
 class StrCategoryConverter(units.ConversionInterface):
     @staticmethod
     def convert(value, unit, axis):
-        """Uses axis.unit_data map to encode
-        data as floats
+        """Uses axis.units to encode string data as floats
+
+        Parameters
+        ----------
+        value: string or iterable
+            value or list of values to be converted
+        unit: None
+            units to use for an axis with string data
+        axis: :class:`~matplotlib.axes.Axes`
+            axis on which the converted value is plotted
         """
-        value = np.atleast_1d(value)
-        # try and update from here....
-        if hasattr(axis.unit_data, 'update'):
-            for val in value:
-                if isinstance(val, six.string_types):
-                    axis.unit_data.update(val)
-        vmap = dict(zip(axis.unit_data.seq, axis.unit_data.locs))
+        # dtype = object preserves numerical pass throughs
+        values = np.atleast_1d(np.array(value, dtype=object))
 
-        if isinstance(value, six.string_types):
-            return vmap[value]
+        # pass through sequence of non binary numbers
+        if all((units.ConversionInterface.is_numlike(v) and
+                not isinstance(v, VALID_TYPES)) for v in values):
+            return np.asarray(values, dtype=float)
 
-        vals = shim_array(value)
+        # force an update so it also does type checking
+        axis.units.update(values)
 
-        for lab, loc in vmap.items():
-            vals[vals == lab] = loc
+        str2idx = np.vectorize(axis.units._mapping.__getitem__,
+                               otypes=[float])
 
-        return vals.astype('float')
+        mapped_value = str2idx(values)
+        return mapped_value
 
     @staticmethod
     def axisinfo(unit, axis):
-        majloc = StrCategoryLocator(axis.unit_data.locs)
-        majfmt = StrCategoryFormatter(axis.unit_data.seq)
+        """Sets the axis ticks and labels
+        """
+        # locator and formatter take mapping dict because
+        # args need to be pass by reference for updates
+        majloc = StrCategoryLocator(axis.units)
+        majfmt = StrCategoryFormatter(axis.units)
         return units.AxisInfo(majloc=majloc, majfmt=majfmt)
 
     @staticmethod
     def default_units(data, axis):
-        # the conversion call stack is:
+        # the conversion call stack is supposed to be
         # default_units->axis_info->convert
-        if axis.unit_data is None:
-            axis.unit_data = UnitData(data)
+        if axis.units is None:
```

Comment:
OK, this needs some explanation, somewhere.  If I use the jpl toy example, `ax.yaxis.units` returns something like 'meters'.  `dates.py` returns the timezone (if set).  Here, you load `axis.units` up with the data map.  I guess thats OK.  But its a little mysterious.  Some description of why would be appreciated.  Is this really the only place we can carry that map around?  

Second, the name `UnitData` doesn't help me know whats going on.  I'd consider changing this name to `CategoricalUnitsMap` or something that makes it clear its categorical that is involved and that its a map.  If I want to query an axis as to its units, `axis.units` is a useful place to look, and `UnitData` doesn't quite convey that (though it will usually be `<matplotlib.category.UnitData object at 0x11eda2d68>`, so take as you will.

Reply:
My understanding is that 'units' is a place to stash what ever the handler feels like stashing there.

Reply:
Yes, although it could do with documenting, I think `units` is a place for the converter to store any variables that affect how it does the conversion.

Reply:
Yes, I see thats how its being used.  But the naive user (me as of a couple of months ago) would have a though time understanding that, and might think that this property of the `axis` class, named`axis.units`, might actually be the units of the axis.

Reply:
With the initial implementation of categorical  we missed this and added `unit_data` (which is new being deprecated) to stash the mapping.

Reply:
the wrinkle to this is that it'd probably be useful to store the sort of sorting info the jpl folks do in their implemention of [StrConvertor](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/testing/jpl_units/StrConverter.py) but that's probably a refactor away.